tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,527 @@
+//!
+//! let response = service
+//!     .ready()
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(StatusCode::NOT_ACCEPTABLE, response.status());
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Custom validation can be made by implementing [`ValidateRequest`]:
+//!
+//! ```
+//! use tower_http::validate_request::{ValidateRequestHeaderLayer, ValidateRequest};
+//! use hyper::{Request, Response, Body, Error};
+//! use http::{StatusCode, header::ACCEPT};
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//!
+//! #[derive(Clone, Copy)]
+//! pub struct MyHeader { /* ...  */ }
+//!
+//! impl<B> ValidateRequest<B> for MyHeader {
+//!     type ResponseBody = Body;
+//!
+//!     fn validate(
+//!         &mut self,
+//!         request: &mut Request<B>,
+//!     ) -> Result<(), Response<Self::ResponseBody>> {
+//!         // validate the request...
+//!         # unimplemented!()
+//!     }
+//! }
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     Ok(Response::new(Body::empty()))
+//! }
+//!
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service = ServiceBuilder::new()
+//!     // Validate requests using `MyHeader`
+//!     .layer(ValidateRequestHeaderLayer::custom(MyHeader { /* ... */ }))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Or using a closure:
+//!
+//! ```
+//! use tower_http::validate_request::{ValidateRequestHeaderLayer, ValidateRequest};
+//! use hyper::{Request, Response, Body, Error};
+//! use http::{StatusCode, header::ACCEPT};
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     # todo!();
+//!     // ...
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service = ServiceBuilder::new()
+//!     .layer(ValidateRequestHeaderLayer::custom(|request: &mut Request<Body>| {
+//!         // Validate the request
+//!         # Ok::<_, Response<Body>>(())
+//!     }))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+
+use http::{header, Request, Response, StatusCode};
+use http_body::Body;
+use mime::Mime;
+use pin_project_lite::pin_project;
+use std::{
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`ValidateRequestHeader`] which validates all requests.
+///
+/// See the [module docs](crate::validate_request) for an example.
+#[derive(Debug, Clone)]
+pub struct ValidateRequestHeaderLayer<T> {
+    validate: T,
+}
+
+impl<ResBody> ValidateRequestHeaderLayer<AcceptHeader<ResBody>> {
+    /// Validate requests have the required Accept header.
+    ///
+    /// The `Accept` header is required to be `*/*`, `type/*` or `type/subtype`,
+    /// as configured.
+    ///
```

Comment:
I think having an example here would be nice otherwise its not clear what `value` should be.

```suggestion
    ///
    /// # Example
    ///
    /// ```
    /// use tower_http::validate_request::ValidateRequestHeaderLayer;
    ///
    /// let layer = ValidateRequestHeaderLayer::accept("application/json");
    /// ```
```